### PR TITLE
DATAREDIS-471 - Add support for partial updates.

### PR DESCRIPTION
**Pull request by christophstrobl**
We now allow partial update of domain types via PartialUpdate. The according expiration times and secondary index structures are updated accordingly.

In some cases it is not necessary to load and rewrite the entire entity just to set a new value within it. A session timestamp for last active time might be such a scenario where you just want to alter one property.
`PartialUpdate` allows to define `set`, `delete` actions on existing objects while taking care of updating potential expiration times of the entity itself as well as index structures.

```java
PartialUpdate<Person> update = new PartialUpdate<Person>("e2c7dcee", Person.class)
  .set("firstname", "mat")                                                           <1>
  .set("address.city", "emond's field")                                              <2>
  .del("age");                                                                       <3>

template.update(update);

update = new PartialUpdate<Person>("e2c7dcee", Person.class)
  .set("address", new Address("caemlyn", "andor"))                                   <4>
  .set("attributes", singletonMap("eye-color", "grey"));                             <5>

template.update(update);

update = new PartialUpdate<Person>("e2c7dcee", Person.class)
  .refreshTtl(true);                                                                 <6>
  .set("expiration", 1000);

template.update(update);
```

```bash
<1> Set the simple property _firstname_ to _mat_
<2> Set the simple property _address.city_ to _emond's field_ without having to pass in the entire object. This does not work when a custom conversion is registered.
<3> Remove the property _age_.
<4> Set complex property _address_.
<5> Set a map/collection of values removes the previously existing map/collection and replaces the values with the given ones.
<6> Automatically update the server expiration time when altering time to live.
```

**NOTE:** Updating complex objects as well as map/collection structures requires further interaction with Redis to determine existing values which means that it might turn out that rewriting the entire entity might be faster.

### CODE REVIEW

File: src/main/java/org/springframework/data/redis/core/PartialUpdate.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	/**
+	 * Remove the value reachable via given {@literal path}.
+	 *
+	 * @param path path must not be {@literal null}.
+	 * @return a new {@link PartialUpdate}.
+	 */
+	public PartialUpdate<T> del(String path) {
+
+		Assert.hasText(path, "Path to remove must not be null or empty!");
+		propertyUpdates.add(new PropertyUpdate(UpdateCommand.DEL, path));
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/core/PartialUpdate.java]: `PartialUpdate` changes its own state and creates a new instance.

File: src/main/java/org/springframework/data/redis/core/convert/MappingRedisConverter.java [hunk 2]
```diff
@@ -357,7 +378,118 @@ public void write(Object source, final RedisData sink) {
...
+
+			String path = pUpdate.getPropertyPath();
+
+			if (UpdateCommand.SET.equals(pUpdate.getCmd())) {
+
+				KeyValuePersistentProperty targetProperty = getTargetPropertyOrNullForPath(path, update.getTarget());
+
+				if (targetProperty == null) {
 
+					writeInternal(entity.getKeySpace(), pUpdate.getPropertyPath(), pUpdate.getValue(),
```
Comment (christophstrobl) [inline src/main/java/org/springframework/data/redis/core/convert/MappingRedisConverter.java]: add a type check and deny eg. setting `foo-bar` as `String` for a `Long` property.

File: src/main/java/org/springframework/data/redis/core/RedisKeyValueAdapter.java [hunk 3]
```diff
@@ -393,6 +394,109 @@ public Long doInRedis(RedisConnection connection) throws DataAccessException {
 		return count != null ? count.longValue() : 0;
 	}
 
+	public void update(final PartialUpdate<?> update) {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/redis/core/RedisKeyValueAdapter.java]: Old index references remain in index-helper index `(entity):(key):idx` when updating/deleting `@Indexed` properties

**Comment by mp911de** [comment]
That's merged with addcdfe.

**Pull request closed without merging**